include/boost/url/impl/decoded_view.ipp

```diff
@@ -0,0 +1,140 @@
+//
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_DECODED_VIEW_IPP
+#define BOOST_URL_IMPL_DECODED_VIEW_IPP
+
+#include <boost/url/decoded_view.hpp>
+#include <ostream>
+
+namespace boost {
+namespace urls {
+
+decoded_view::
+decoded_view(
+    string_view str,
+    pct_decode_opts opt) noexcept
+    : p_(str.data())
+    , n_(str.size())
+    , plus_to_space_(opt.plus_to_space)
+{
+    error_code ec;
+    opt.non_normal_is_error = false;
+    dn_ = validate_pct_encoding(
+        str, ec, opt,
+        [](unsigned char c) { return c != '%'; });
+    if (ec.failed())
+        detail::throw_invalid_argument(
+            BOOST_CURRENT_LOCATION);
+}
+
+decoded_view::size_type
+decoded_view::
+copy(
+    char* dest,
+    size_type count,
+    size_type pos) const
+{
+    if( pos > size() )
+        detail::throw_invalid_argument(
+            BOOST_CURRENT_LOCATION);
+    std::size_t rlen = (std::min)(count, size() - pos);
+    auto first = begin();
+    while (pos--)
+        ++first;
+    auto last = first;
+    while (rlen--)
+        ++last;
+    for (; first != last; ++first, ++dest)
+        *dest = *first;
+    return rlen;
+}
+
+int
+decoded_view::
+compare(string_view other) const noexcept
+{
+    std::size_t rlen = (std::min)(size(), other.size());
+
+    // int cmp = traits_type::compare(encoded_data(), other.data(), rlen);
+    auto first1 = begin();
+    auto last1 = end();
+    auto first2 = other.begin();
+    auto last2 = other.end();
+    int cmp = 0;
+    for ( ;
+         (first1 != last1) && (first2 != last2);
+         ++first1, ++first2, --rlen)
+    {
+        if (*first1 < *first2)
+        {
+            cmp = -1;
+            break;
+        }
+        else if (*first2 < *first1)
+        {
+            cmp = +1;
+            break;
+        }
+    }
```

Comment:
This loop has some pretty significant inefficiencies, can you find them?

Reply:
I improved it a little, but you probably want to check it again.

Reply:
Each iterator is being checked THREE times for a percent. And the `while` can be simplified to use a counter, because we know in advance if one decoded string is shorter. Furthermore, when we calculate that one of the strings is shorter, we can use that information to avoid the additional conditions on lines 97 and 99. Consult a generic implementation of 3-way compare for tips.

Reply:
I had a look at `strcmp` and tried to improve it a little more.

Reply:
Yeah its better now! Still can do better though :)